Decompiler: Support more 64-bit Linux hosts

zeldin opened the pull request:
This adds support for building the decompiler on more 64-bit Linux hosts, namely Linux/ppc64 (big and little endian), Linux/arm64 (big and little endian) and Linux/riscv.

nihilus commented:
There already was decent support, did you try building with "-m32" ?

zeldin commented:
@nihilus This issue is specifically about 64-bit support.  I have a 64-bit only userland.  :-)
That being said, it doesn't look like Linux/ppc32 is supported either, just MacOS.

nihilus commented:
@zeldin maybe you have a mips64 host running Linux, a pity my are still only running IRIX and are stuffed up in a cellar.

Funny enough I just realized by github's email notification that we used to work together at 27M.

zeldin commented:
@nihilus I'm afraid I don't have any mips64 systems, and no alpha or ia64 ones either.  (And 64-bit PA-RISC I don't even think I've come into contact with.) I _do_ have a couple of 64-bit SPARCs, but none of them is running Linux at the moment.  :smile: 
Alas 27M is no more...  :cry:

nihilus commented:
Yeah, I heard you guys were merged Zenterio.

I was thinking like this about the pull request; wouldn't it make more sense to typedef int2 to int16_t and use types.h/cc for getting architecture specific details?

zeldin commented:
Yeah, relying on `<stdint.h>` could remove all the conditional typedefs in one fell swoop.  That would still leave the HOST_ENDIAN definition though.  On Linux specifically you can just use `<endian.h>`, dunno about MacOS, the other operating systems listed probably only run on one endianness each.
Of course, since this is nominally C++ and not C, one could potentially use std::endian.  Apart from being C++20, it wouldn't interact with the preprocessor like a define does though...

nihilus commented:
You could make it like this '#define big_endian (nthol(0xdeadbeef) == 0xdeadbeef)'

However for the case of mixed endian one must check if it both non-little endian and non-big enndian.

zeldin commented:
I rebased the decompiler fixes to current master, and also added a second commit with fixes to the build system.  Building has been tested on ppc64 and arm64.

nullterminated commented:
zeldin:linux64 builds and runs on my machine too. I'll close PR#572 as this now supersedes it.

ryanmkurtz commented:
I am going to close this for now since the Ghidra team cannot properly support and maintain builds for other architectures at this time.  See [this discussion for more info](https://github.com/NationalSecurityAgency/ghidra/pull/730#issuecomment-511946890).  If something changes in the future, we will use this as a reference.  Thanks!

zeldin commented:
I'm not asking you to support or maintain builds, just to merge this to make things easier for downstreams maintainers making and/or supporting such builds...

ryanmkurtz commented:
I understand, but if we merge it in and we can't test it, we can't know if we broke it if we need to refactor.  Our build tends to get refactored a lot.

kant2002 commented:
@zeldin if you want, you may send PR to my fork. I would try to keep small infrastructure changes in Ghidra there. Changes like this one which Ghidra cannot support right now.

zeldin commented:
@ryanmkurtz Well, if you are refactoring the build system a lot, might I suggest that at the next refactoring you create a build target for just building natively with the default compiler flags without worrying about any "platform name"?  You could test _that_ at least...

Manouchehri commented:
Thanks for the patch, works great here!

In case anyone is interested, here's my build commands for a Ubuntu 19.04 ARM64 host:

```sh
git clone https://github.com/NationalSecurityAgency/ghidra.git
cd ghidra
git remote add zeldin https://github.com/zeldin/ghidra.git
git fetch zeldin linux64
git cherry-pick 9bce711 6a613c9

sudo add-apt-repository ppa:cwchien/gradle
sudo apt-get update
sudo apt-get install openjdk-13-jdk gradle clang-8
sudo update-alternatives --install /usr/bin/clang clang /usr/bin/clang-8 100
sudo update-alternatives --install /usr/bin/clang++ clang++ /usr/bin/clang++-8 100

gradle --init-script gradle/support/fetchDependencies.gradle init
gradle buildGhidra
```
